// html-document/lib-cov/Document.js

// instrument by jscoverage, do not modifly this file
(function (file, lines, conds, source) {
  var BASE;
  if (typeof global === 'object') {
    BASE = global;
  } else if (typeof window === 'object') {
    BASE = window;
  } else {
    throw new Error('[jscoverage] unknow ENV!');
  }
  if (BASE._$jscoverage) {
    BASE._$jscmd(file, 'init', lines, conds, source);
    return;
  }
  var cov = {};
  /**
   * jsc(file, 'init', lines, condtions)
   * jsc(file, 'line', lineNum)
   * jsc(file, 'cond', lineNum, expr, start, offset)
   */
  function jscmd(file, type, line, express, start, offset) {
    var storage;
    switch (type) {
      case 'init':
        if(cov[file]){
          storage = cov[file];
        } else {
          storage = [];
          for (var i = 0; i < line.length; i ++) {
            storage[line[i]] = 0;
          }
          var condition = express;
          var source = start;
          storage.condition = condition;
          storage.source = source;
        }
        cov[file] = storage;
        break;
      case 'line':
        storage = cov[file];
        storage[line] ++;
        break;
      case 'cond':
        storage = cov[file];
        storage.condition[line] ++;
        return express;
    }
  }

  BASE._$jscoverage = cov;
  BASE._$jscmd = jscmd;
  jscmd(file, 'init', lines, conds, source);
})('lib/Document.js', [1,3,15,16,17,18,19,22,23,25,94,97,4,12,26,34,36,42,48,65,71,75,79,83,87,91,27,28,29,30,31,37,38,39,43,44,45,49,61,62,52,53,55,56,58,59,66,67,68,72,76,80,84,88], {}, ["\"use strict\";","","var _extends = function (child, parent) {","  child.prototype = Object.create(parent.prototype, {","    constructor: {","      value: child,","      enumerable: false,","      writable: true,","      configurable: true","    }","  });","  child.__proto__ = parent;","};","","var Node = require(\"./Node\").Node;","var Comment = require(\"./Comment\").Comment;","var DocumentFragment = require(\"./DocumentFragment\").DocumentFragment;","var HTMLElement = require(\"./HTMLElement\").HTMLElement;","var Text = require(\"./Text\").Text;","","// HTML Elements","var HTMLOptionElement = require(\"./HTMLElement/elements/HTMLOptionElement\").HTMLOptionElement;","var HTMLSelectElement = require(\"./HTMLElement/elements/HTMLSelectElement\").HTMLSelectElement;","","var Document = (function (Node) {","  var Document = function Document() {","    this.documentElement = this.createElement(\"html\");","    this.head = this.createElement(\"head\");","    this.documentElement.appendChild(this.head);","    this.body = this.createElement(\"body\");","    this.documentElement.appendChild(this.body);","  };","","  _extends(Document, Node);","","  Document.prototype.createComment = function (data /* : String*/) {","    var comment = new Comment(data);","    comment.ownerDocument = this;","    return comment;","  };","","  Document.prototype.createDocumentFragment = function () {","    var fragment = new DocumentFragment();","    fragment.ownerDocument = this;","    return fragment;","  };","","  Document.prototype.createElement = function (name) {","    var element;","    switch (name.toLowerCase()) {","      case \"select\":","        element = new HTMLSelectElement();","        break;","      case \"option\":","        element = new HTMLOptionElement();","        break;","      default:","        element = new HTMLElement();","        element.nodeName = name;","    }","    element.ownerDocument = this;","    return element;","  };","","  Document.prototype.createTextNode = function (textContent) {","    var text = new Text(textContent);","    text.ownerDocument = this;","    return text;","  };","","  Document.prototype.getElementById = function (id) {","    return this.documentElement.getElementById(id);","  };","","  Document.prototype.getElementsByTagName = function (tagName, _array) {","    return this.documentElement.getElementsByTagName(tagName, _array);","  };","","  Document.prototype.getElementsByClassName = function (className) {","    throw new Error(\"Not implemented\");","  };","","  Document.prototype.querySelector = function () {","    throw new Error(\"Not implemented\");","  };","","  Document.prototype.querySelectorAll = function () {","    throw new Error(\"Not implemented\");","  };","","  return Document;","})(Node);","","exports.Document = Document;","","","Object.defineProperty(Document.prototype, \"nodeType\", { value: Node.DOCUMENT_NODE });","//# sourceMappingURL=Document.js.map"]);
_$jscmd("lib/Document.js", "line", 1);

"use strict";

_$jscmd("lib/Document.js", "line", 3);

var _extends = function(child, parent) {
    _$jscmd("lib/Document.js", "line", 4);
    child.prototype = Object.create(parent.prototype, {
        constructor: {
            value: child,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    _$jscmd("lib/Document.js", "line", 12);
    child.__proto__ = parent;
};

_$jscmd("lib/Document.js", "line", 15);

var Node = require("./Node").Node;

_$jscmd("lib/Document.js", "line", 16);

var Comment = require("./Comment").Comment;

_$jscmd("lib/Document.js", "line", 17);

var DocumentFragment = require("./DocumentFragment").DocumentFragment;

_$jscmd("lib/Document.js", "line", 18);

var HTMLElement = require("./HTMLElement").HTMLElement;

_$jscmd("lib/Document.js", "line", 19);

var Text = require("./Text").Text;

_$jscmd("lib/Document.js", "line", 22);

// HTML Elements
var HTMLOptionElement = require("./HTMLElement/elements/HTMLOptionElement").HTMLOptionElement;

_$jscmd("lib/Document.js", "line", 23);

var HTMLSelectElement = require("./HTMLElement/elements/HTMLSelectElement").HTMLSelectElement;

_$jscmd("lib/Document.js", "line", 25);

var Document = function(Node) {
    _$jscmd("lib/Document.js", "line", 26);
    var Document = function Document() {
        _$jscmd("lib/Document.js", "line", 27);
        this.documentElement = this.createElement("html");
        _$jscmd("lib/Document.js", "line", 28);
        this.head = this.createElement("head");
        _$jscmd("lib/Document.js", "line", 29);
        this.documentElement.appendChild(this.head);
        _$jscmd("lib/Document.js", "line", 30);
        this.body = this.createElement("body");
        _$jscmd("lib/Document.js", "line", 31);
        this.documentElement.appendChild(this.body);
    };
    _$jscmd("lib/Document.js", "line", 34);
    _extends(Document, Node);
    _$jscmd("lib/Document.js", "line", 36);
    Document.prototype.createComment = function(data) {
        _$jscmd("lib/Document.js", "line", 37);
        var comment = new Comment(data);
        _$jscmd("lib/Document.js", "line", 38);
        comment.ownerDocument = this;
        _$jscmd("lib/Document.js", "line", 39);
        return comment;
    };
    _$jscmd("lib/Document.js", "line", 42);
    Document.prototype.createDocumentFragment = function() {
        _$jscmd("lib/Document.js", "line", 43);
        var fragment = new DocumentFragment();
        _$jscmd("lib/Document.js", "line", 44);
        fragment.ownerDocument = this;
        _$jscmd("lib/Document.js", "line", 45);
        return fragment;
    };
    _$jscmd("lib/Document.js", "line", 48);
    Document.prototype.createElement = function(name) {
        _$jscmd("lib/Document.js", "line", 49);
        var element;
        switch (name.toLowerCase()) {
          case "select":
            _$jscmd("lib/Document.js", "line", 52);
            element = new HTMLSelectElement();
            _$jscmd("lib/Document.js", "line", 53);
            break;

          case "option":
            _$jscmd("lib/Document.js", "line", 55);
            element = new HTMLOptionElement();
            _$jscmd("lib/Document.js", "line", 56);
            break;

          default:
            _$jscmd("lib/Document.js", "line", 58);
            element = new HTMLElement();
            _$jscmd("lib/Document.js", "line", 59);
            element.nodeName = name;
        }
        _$jscmd("lib/Document.js", "line", 61);
        element.ownerDocument = this;
        _$jscmd("lib/Document.js", "line", 62);
        return element;
    };
    _$jscmd("lib/Document.js", "line", 65);
    Document.prototype.createTextNode = function(textContent) {
        _$jscmd("lib/Document.js", "line", 66);
        var text = new Text(textContent);
        _$jscmd("lib/Document.js", "line", 67);
        text.ownerDocument = this;
        _$jscmd("lib/Document.js", "line", 68);
        return text;
    };
    _$jscmd("lib/Document.js", "line", 71);
    Document.prototype.getElementById = function(id) {
        _$jscmd("lib/Document.js", "line", 72);
        return this.documentElement.getElementById(id);
    };
    _$jscmd("lib/Document.js", "line", 75);
    Document.prototype.getElementsByTagName = function(tagName, _array) {
        _$jscmd("lib/Document.js", "line", 76);
        return this.documentElement.getElementsByTagName(tagName, _array);
    };
    _$jscmd("lib/Document.js", "line", 79);
    Document.prototype.getElementsByClassName = function(className) {
        _$jscmd("lib/Document.js", "line", 80);
        throw new Error("Not implemented");
    };
    _$jscmd("lib/Document.js", "line", 83);
    Document.prototype.querySelector = function() {
        _$jscmd("lib/Document.js", "line", 84);
        throw new Error("Not implemented");
    };
    _$jscmd("lib/Document.js", "line", 87);
    Document.prototype.querySelectorAll = function() {
        _$jscmd("lib/Document.js", "line", 88);
        throw new Error("Not implemented");
    };
    _$jscmd("lib/Document.js", "line", 91);
    return Document;
}(Node);

_$jscmd("lib/Document.js", "line", 94);

exports.Document = Document;

_$jscmd("lib/Document.js", "line", 97);

Object.defineProperty(Document.prototype, "nodeType", {
    value: Node.DOCUMENT_NODE
});
