// html-document/lib-cov/Element.js

// instrument by jscoverage, do not modifly this file
(function (file, lines, conds, source) {
  var BASE;
  if (typeof global === 'object') {
    BASE = global;
  } else if (typeof window === 'object') {
    BASE = window;
  } else {
    throw new Error('[jscoverage] unknow ENV!');
  }
  if (BASE._$jscoverage) {
    BASE._$jscmd(file, 'init', lines, conds, source);
    return;
  }
  var cov = {};
  /**
   * jsc(file, 'init', lines, condtions)
   * jsc(file, 'line', lineNum)
   * jsc(file, 'cond', lineNum, expr, start, offset)
   */
  function jscmd(file, type, line, express, start, offset) {
    var storage;
    switch (type) {
      case 'init':
        if(cov[file]){
          storage = cov[file];
        } else {
          storage = [];
          for (var i = 0; i < line.length; i ++) {
            storage[line[i]] = 0;
          }
          var condition = express;
          var source = start;
          storage.condition = condition;
          storage.source = source;
        }
        cov[file] = storage;
        break;
      case 'line':
        storage = cov[file];
        storage[line] ++;
        break;
      case 'cond':
        storage = cov[file];
        storage.condition[line] ++;
        return express;
    }
  }

  BASE._$jscoverage = cov;
  BASE._$jscmd = jscmd;
  jscmd(file, 'init', lines, conds, source);
})('lib/Element.js', [1,3,8,20,21,23,133,136,9,17,24,28,30,38,52,130,25,31,33,42,43,44,49,40,46,60,70,80,93,94,105,115,125], {"4_6_11":0,"5_6_13":0,"31_11_145":0,"32_10_51":0,"32_10_20":0,"32_34_27":0,"39_8_8":0,"40_23_21":0,"40_47_40":0,"42_13_6":0,"42_23_2":0,"45_10_36":0,"105_15_19":0,"115_47_4":0,"115_54_45":0}, ["\"use strict\";","","var _classProps = function (child, staticProps, instanceProps) {","  if (staticProps) Object.defineProperties(child, staticProps);","  if (instanceProps) Object.defineProperties(child.prototype, instanceProps);","};","","var _extends = function (child, parent) {","  child.prototype = Object.create(parent.prototype, {","    constructor: {","      value: child,","      enumerable: false,","      writable: true,","      configurable: true","    }","  });","  child.__proto__ = parent;","};","","var Node = require(\"./Node\").Node;","var ParentNode = require(\"./ParentNode\").ParentNode;","","var Element = (function (ParentNode) {","  var Element = function Element() {","    ParentNode.apply(this, arguments);","  };","","  _extends(Element, ParentNode);","","  Element.prototype.getElementById = function (id) {","    return this._childNodesRecursiveFind(function (e) {","      if (e instanceof Element && e.getAttribute(\"id\") === id) {","        return true;","      }","    }) || null;","  };","","  Element.prototype.getElementsByTagName = function (tagName, _array) {","    if (!tagName) {","      return !_array ? this.children.slice() : _array.push.apply(_array, this.children);","    }","    _array = _array || [];","    tagName = tagName.toLowerCase();","    this.children.forEach(function (e) {","      if (e.nodeName.toLowerCase() === tagName) {","        _array.push(e);","      }","    });","    return _array;","  };","","  _classProps(Element, null, {","    id: {","      /**","       * Gets the id of the element.","       *","       * @return {String}","       */","      get: function () {","        return this.getAttribute(\"id\");","      },","","","      /**","       * Sets the id of the element.","       *","       * @param {String} id","       */","      set: function (id) {","        this.setAttribute(\"id\", id);","      }","    },","    tagName: {","      /**","       * Gets the tagName of the element.","       *","       * @return {String}","       */","      get: function () {","        return this.nodeName;","      }","    },","    children: {","      /**","       * Returns a live {@link HTMLCollection} containing all objects of type {@link Element}","       * that are children of this ParentNode.","       *","       * Note: this currently returns a non-live array.","       *","       * @return {HTMLCollection}","       */","      get: function () {","        return this._childNodes.filter(function (n) {","          return n instanceof Element;","        });","      }","    },","    firstElementChild: {","      /**","       * Returns the {@link Element} that is the first child of this ParentNode, or null if there is none.","       *","       * @return {Element}","       */","      get: function () {","        return this._childNodes[0] || null;","      }","    },","    lastElementChild: {","      /**","       * Returns the {@link Element} that is the first child of this ParentNode, or null if there is none.","       *","       * @return {Element}","       */","      get: function () {","        return this._childNodes.length === 0 ? null : this._childNodes[this._childNodes.length - 1];","      }","    },","    childElementCount: {","      /**","       * Returns an unsigned long giving the amount of children that the object has.","       *","       * @return {Number}","       */","      get: function () {","        return this._childNodes.length;","      }","    }","  });","","  return Element;","})(ParentNode);","","exports.Element = Element;","","","Object.defineProperty(Element.prototype, \"nodeType\", { value: Node.ELEMENT_NODE });","//# sourceMappingURL=Element.js.map"]);
_$jscmd("lib/Element.js", "line", 1);

"use strict";

_$jscmd("lib/Element.js", "line", 3);

var _classProps = function(child, staticProps, instanceProps) {
    if (_$jscmd("lib/Element.js", "cond", "4_6_11", staticProps)) Object.defineProperties(child, staticProps);
    if (_$jscmd("lib/Element.js", "cond", "5_6_13", instanceProps)) Object.defineProperties(child.prototype, instanceProps);
};

_$jscmd("lib/Element.js", "line", 8);

var _extends = function(child, parent) {
    _$jscmd("lib/Element.js", "line", 9);
    child.prototype = Object.create(parent.prototype, {
        constructor: {
            value: child,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    _$jscmd("lib/Element.js", "line", 17);
    child.__proto__ = parent;
};

_$jscmd("lib/Element.js", "line", 20);

var Node = require("./Node").Node;

_$jscmd("lib/Element.js", "line", 21);

var ParentNode = require("./ParentNode").ParentNode;

_$jscmd("lib/Element.js", "line", 23);

var Element = function(ParentNode) {
    _$jscmd("lib/Element.js", "line", 24);
    var Element = function Element() {
        _$jscmd("lib/Element.js", "line", 25);
        ParentNode.apply(this, arguments);
    };
    _$jscmd("lib/Element.js", "line", 28);
    _extends(Element, ParentNode);
    _$jscmd("lib/Element.js", "line", 30);
    Element.prototype.getElementById = function(id) {
        _$jscmd("lib/Element.js", "line", 31);
        return _$jscmd("lib/Element.js", "cond", "31_11_145", this._childNodesRecursiveFind(function(e) {
            if (_$jscmd("lib/Element.js", "cond", "32_10_51", _$jscmd("lib/Element.js", "cond", "32_10_20", e instanceof Element) && _$jscmd("lib/Element.js", "cond", "32_34_27", e.getAttribute("id") === id))) {
                _$jscmd("lib/Element.js", "line", 33);
                return true;
            }
        })) || null;
    };
    _$jscmd("lib/Element.js", "line", 38);
    Element.prototype.getElementsByTagName = function(tagName, _array) {
        if (_$jscmd("lib/Element.js", "cond", "39_8_8", !tagName)) {
            _$jscmd("lib/Element.js", "line", 40);
            return !_array ? _$jscmd("lib/Element.js", "cond", "40_23_21", this.children.slice()) : _$jscmd("lib/Element.js", "cond", "40_47_40", _array.push.apply(_array, this.children));
        }
        _$jscmd("lib/Element.js", "line", 42);
        _array = _$jscmd("lib/Element.js", "cond", "42_13_6", _array) || _$jscmd("lib/Element.js", "cond", "42_23_2", []);
        _$jscmd("lib/Element.js", "line", 43);
        tagName = tagName.toLowerCase();
        _$jscmd("lib/Element.js", "line", 44);
        this.children.forEach(function(e) {
            if (_$jscmd("lib/Element.js", "cond", "45_10_36", e.nodeName.toLowerCase() === tagName)) {
                _$jscmd("lib/Element.js", "line", 46);
                _array.push(e);
            }
        });
        _$jscmd("lib/Element.js", "line", 49);
        return _array;
    };
    _$jscmd("lib/Element.js", "line", 52);
    _classProps(Element, null, {
        id: {
            /**
       * Gets the id of the element.
       *
       * @return {String}
       */
            get: function() {
                _$jscmd("lib/Element.js", "line", 60);
                return this.getAttribute("id");
            },
            /**
       * Sets the id of the element.
       *
       * @param {String} id
       */
            set: function(id) {
                _$jscmd("lib/Element.js", "line", 70);
                this.setAttribute("id", id);
            }
        },
        tagName: {
            /**
       * Gets the tagName of the element.
       *
       * @return {String}
       */
            get: function() {
                _$jscmd("lib/Element.js", "line", 80);
                return this.nodeName;
            }
        },
        children: {
            /**
       * Returns a live {@link HTMLCollection} containing all objects of type {@link Element}
       * that are children of this ParentNode.
       *
       * Note: this currently returns a non-live array.
       *
       * @return {HTMLCollection}
       */
            get: function() {
                _$jscmd("lib/Element.js", "line", 93);
                return this._childNodes.filter(function(n) {
                    _$jscmd("lib/Element.js", "line", 94);
                    return n instanceof Element;
                });
            }
        },
        firstElementChild: {
            /**
       * Returns the {@link Element} that is the first child of this ParentNode, or null if there is none.
       *
       * @return {Element}
       */
            get: function() {
                _$jscmd("lib/Element.js", "line", 105);
                return _$jscmd("lib/Element.js", "cond", "105_15_19", this._childNodes[0]) || null;
            }
        },
        lastElementChild: {
            /**
       * Returns the {@link Element} that is the first child of this ParentNode, or null if there is none.
       *
       * @return {Element}
       */
            get: function() {
                _$jscmd("lib/Element.js", "line", 115);
                return this._childNodes.length === 0 ? _$jscmd("lib/Element.js", "cond", "115_47_4", null) : _$jscmd("lib/Element.js", "cond", "115_54_45", this._childNodes[this._childNodes.length - 1]);
            }
        },
        childElementCount: {
            /**
       * Returns an unsigned long giving the amount of children that the object has.
       *
       * @return {Number}
       */
            get: function() {
                _$jscmd("lib/Element.js", "line", 125);
                return this._childNodes.length;
            }
        }
    });
    _$jscmd("lib/Element.js", "line", 130);
    return Element;
}(ParentNode);

_$jscmd("lib/Element.js", "line", 133);

exports.Element = Element;

_$jscmd("lib/Element.js", "line", 136);

Object.defineProperty(Element.prototype, "nodeType", {
    value: Node.ELEMENT_NODE
});
