// html-document/lib-cov/Node.js

// instrument by jscoverage, do not modifly this file
(function (file, lines, conds, source) {
  var BASE;
  if (typeof global === 'object') {
    BASE = global;
  } else if (typeof window === 'object') {
    BASE = window;
  } else {
    throw new Error('[jscoverage] unknow ENV!');
  }
  if (BASE._$jscoverage) {
    BASE._$jscmd(file, 'init', lines, conds, source);
    return;
  }
  var cov = {};
  /**
   * jsc(file, 'init', lines, condtions)
   * jsc(file, 'line', lineNum)
   * jsc(file, 'cond', lineNum, expr, start, offset)
   */
  function jscmd(file, type, line, express, start, offset) {
    var storage;
    switch (type) {
      case 'init':
        if(cov[file]){
          storage = cov[file];
        } else {
          storage = [];
          for (var i = 0; i < line.length; i ++) {
            storage[line[i]] = 0;
          }
          var condition = express;
          var source = start;
          storage.condition = condition;
          storage.source = source;
        }
        cov[file] = storage;
        break;
      case 'line':
        storage = cov[file];
        storage[line] ++;
        break;
      case 'cond':
        storage = cov[file];
        storage.condition[line] ++;
        return express;
    }
  }

  BASE._$jscoverage = cov;
  BASE._$jscmd = jscmd;
  jscmd(file, 'init', lines, conds, source);
})('lib/Node.js', [1,3,8,120,123,124,125,126,127,128,129,130,131,132,133,134,9,13,17,22,26,31,33,37,52,64,104,117,10,14,18,19,23,27,28,34,38,42,48,49,40,44,46,53,55,56,60,58,65,67,82,101,74,69,72,76,77,78,86,98,87,89,90,91,95,112], {"4_6_11":0,"5_6_13":0,"38_38_23":0,"38_64_22":0,"39_8_17":0,"43_8_32":0,"53_38_23":0,"53_64_22":0,"54_8_51":0,"54_8_16":0,"54_28_31":0,"57_10_8":0,"68_10_16":0,"71_10_24":0,"74_22_26":0,"74_52_42":0,"75_10_9":0,"85_8_25":0,"87_24_25":0,"87_53_41":0,"88_12_9":0,"94_12_45":0,"94_12_25":0,"94_41_16":0}, ["\"use strict\";","","var _classProps = function (child, staticProps, instanceProps) {","  if (staticProps) Object.defineProperties(child, staticProps);","  if (instanceProps) Object.defineProperties(child.prototype, instanceProps);","};","","var Node = (function () {","  var Node = function Node() {","    this._attributes = {};","  };","","  Node.prototype.getAttribute = function (attributeName) {","    return this._attributes[attributeName];","  };","","  Node.prototype.setAttribute = function (attributeName, attributeValue) {","    this._setAttribute(attributeName, attributeValue);","    this._updatedAttribute(attributeName, this._attributes[attributeName]);","  };","","  Node.prototype._setAttribute = function (attributeName, attributeValue) {","    this._attributes[attributeName] = String(attributeValue);","  };","","  Node.prototype.removeAttribute = function (attributeName) {","    delete this._attributes[attributeName];","    this._updatedAttribute(attributeName);","  };","","  Node.prototype._updatedAttribute = function (attributeName) {};","","  Node.prototype._toHTML = function () {","    return this.outerHTML;","  };","","  Node.prototype.addEventListener = function (eventType, listener, capturingPhase) {","    var _eventsKey = capturingPhase ? \"_eventsCapturingPhase\" : \"_eventsBubblingPhase\";","    if (!this[_eventsKey]) {","      this[_eventsKey] = new Map();","    }","    var callbacks;","    if (!this[_eventsKey].has(eventType)) {","      this[_eventsKey].set(eventType, callbacks = []);","    } else {","      callbacks = this[_eventsKey].get(eventType);","    }","    callbacks.push(listener);","    return this;","  };","","  Node.prototype.removeEventListener = function (eventType, listener, capturingPhase) {","    var _eventsKey = capturingPhase ? \"_eventsCapturingPhase\" : \"_eventsBubblingPhase\";","    if (this[_eventsKey] && this[_eventsKey].has(eventType)) {","      var callbacks = this[_eventsKey].get(eventType);","      var i = callbacks.indexOf(listener);","      if (i === -1) {","        return false;","      }","      callbacks.splice(i, 1);","    }","  };","","  Node.prototype.dispatchEvent = function (event) {","    event.target = this;","    // Capturing phase","    var capturingPhase = function (event) {","      if (this._parentNode) {","        capturingPhase.call(this._parentNode, event);","      }","      if (event.propagationStopped) {","        return;","      }","      var callbacks = this._eventsCapturingPhase && this._eventsCapturingPhase.get(event.type);","      if (callbacks) {","        callbacks.some(function (callback) {","          callback();","          return event.immediatePropagationStopped;","        });","      }","    };","    capturingPhase.call(this, event);","","    // Bubbling phase","    if (!event.propagationStopped) {","      var bubblingPhase = function (event) {","        var callbacks = this._eventsBubblingPhase && this._eventsBubblingPhase.get(event.type);","        if (callbacks) {","          callbacks.some(function (callback) {","            callback();","            return event.immediatePropagationStopped;","          });","        }","        if (!event.propagationStopped && this._parentNode) {","          bubblingPhase.call(this._parentNode, event);","        }","      };","      bubblingPhase.call(this, event);","    }","","    return !event.defaultPrevented;","  };","","  _classProps(Node, null, {","    textContent: {","      /* CONTENT */","","      /**","       * @return {String}","       */","      get: function () {","        return \"\";","      }","    }","  });","","  return Node;","})();","","exports.Node = Node;","","","Object.defineProperty(Node, \"ELEMENT_NODE\", { value: 1 });","Object.defineProperty(Node, \"ATTRIBUTE_NODE\", { value: 2 });","Object.defineProperty(Node, \"TEXT_NODE\", { value: 3 });","Object.defineProperty(Node, \"CDATA_SECTION_NODE\", { value: 4 });","Object.defineProperty(Node, \"ENTITY_REFERENCE_NODE\", { value: 5 });","Object.defineProperty(Node, \"ENTITY_NODE\", { value: 6 });","Object.defineProperty(Node, \"PROCESSING_INSTRUCTION_NODE\", { value: 7 });","Object.defineProperty(Node, \"COMMENT_NODE\", { value: 8 });","Object.defineProperty(Node, \"DOCUMENT_NODE\", { value: 9 });","Object.defineProperty(Node, \"DOCUMENT_TYPE_NODE\", { value: 10 });","Object.defineProperty(Node, \"DOCUMENT_FRAGMENT_NODE\", { value: 11 });","Object.defineProperty(Node, \"NOTATION_NODE\", { value: 12 });","//# sourceMappingURL=Node.js.map"]);
_$jscmd("lib/Node.js", "line", 1);

"use strict";

_$jscmd("lib/Node.js", "line", 3);

var _classProps = function(child, staticProps, instanceProps) {
    if (_$jscmd("lib/Node.js", "cond", "4_6_11", staticProps)) Object.defineProperties(child, staticProps);
    if (_$jscmd("lib/Node.js", "cond", "5_6_13", instanceProps)) Object.defineProperties(child.prototype, instanceProps);
};

_$jscmd("lib/Node.js", "line", 8);

var Node = function() {
    _$jscmd("lib/Node.js", "line", 9);
    var Node = function Node() {
        _$jscmd("lib/Node.js", "line", 10);
        this._attributes = {};
    };
    _$jscmd("lib/Node.js", "line", 13);
    Node.prototype.getAttribute = function(attributeName) {
        _$jscmd("lib/Node.js", "line", 14);
        return this._attributes[attributeName];
    };
    _$jscmd("lib/Node.js", "line", 17);
    Node.prototype.setAttribute = function(attributeName, attributeValue) {
        _$jscmd("lib/Node.js", "line", 18);
        this._setAttribute(attributeName, attributeValue);
        _$jscmd("lib/Node.js", "line", 19);
        this._updatedAttribute(attributeName, this._attributes[attributeName]);
    };
    _$jscmd("lib/Node.js", "line", 22);
    Node.prototype._setAttribute = function(attributeName, attributeValue) {
        _$jscmd("lib/Node.js", "line", 23);
        this._attributes[attributeName] = String(attributeValue);
    };
    _$jscmd("lib/Node.js", "line", 26);
    Node.prototype.removeAttribute = function(attributeName) {
        _$jscmd("lib/Node.js", "line", 27);
        delete this._attributes[attributeName];
        _$jscmd("lib/Node.js", "line", 28);
        this._updatedAttribute(attributeName);
    };
    _$jscmd("lib/Node.js", "line", 31);
    Node.prototype._updatedAttribute = function(attributeName) {};
    _$jscmd("lib/Node.js", "line", 33);
    Node.prototype._toHTML = function() {
        _$jscmd("lib/Node.js", "line", 34);
        return this.outerHTML;
    };
    _$jscmd("lib/Node.js", "line", 37);
    Node.prototype.addEventListener = function(eventType, listener, capturingPhase) {
        _$jscmd("lib/Node.js", "line", 38);
        var _eventsKey = capturingPhase ? _$jscmd("lib/Node.js", "cond", "38_38_23", "_eventsCapturingPhase") : _$jscmd("lib/Node.js", "cond", "38_64_22", "_eventsBubblingPhase");
        if (_$jscmd("lib/Node.js", "cond", "39_8_17", !this[_eventsKey])) {
            _$jscmd("lib/Node.js", "line", 40);
            this[_eventsKey] = new Map();
        }
        _$jscmd("lib/Node.js", "line", 42);
        var callbacks;
        if (_$jscmd("lib/Node.js", "cond", "43_8_32", !this[_eventsKey].has(eventType))) {
            _$jscmd("lib/Node.js", "line", 44);
            this[_eventsKey].set(eventType, callbacks = []);
        } else {
            _$jscmd("lib/Node.js", "line", 46);
            callbacks = this[_eventsKey].get(eventType);
        }
        _$jscmd("lib/Node.js", "line", 48);
        callbacks.push(listener);
        _$jscmd("lib/Node.js", "line", 49);
        return this;
    };
    _$jscmd("lib/Node.js", "line", 52);
    Node.prototype.removeEventListener = function(eventType, listener, capturingPhase) {
        _$jscmd("lib/Node.js", "line", 53);
        var _eventsKey = capturingPhase ? _$jscmd("lib/Node.js", "cond", "53_38_23", "_eventsCapturingPhase") : _$jscmd("lib/Node.js", "cond", "53_64_22", "_eventsBubblingPhase");
        if (_$jscmd("lib/Node.js", "cond", "54_8_51", _$jscmd("lib/Node.js", "cond", "54_8_16", this[_eventsKey]) && _$jscmd("lib/Node.js", "cond", "54_28_31", this[_eventsKey].has(eventType)))) {
            _$jscmd("lib/Node.js", "line", 55);
            var callbacks = this[_eventsKey].get(eventType);
            _$jscmd("lib/Node.js", "line", 56);
            var i = callbacks.indexOf(listener);
            if (_$jscmd("lib/Node.js", "cond", "57_10_8", i === -1)) {
                _$jscmd("lib/Node.js", "line", 58);
                return false;
            }
            _$jscmd("lib/Node.js", "line", 60);
            callbacks.splice(i, 1);
        }
    };
    _$jscmd("lib/Node.js", "line", 64);
    Node.prototype.dispatchEvent = function(event) {
        _$jscmd("lib/Node.js", "line", 65);
        event.target = this;
        _$jscmd("lib/Node.js", "line", 67);
        // Capturing phase
        var capturingPhase = function(event) {
            if (_$jscmd("lib/Node.js", "cond", "68_10_16", this._parentNode)) {
                _$jscmd("lib/Node.js", "line", 69);
                capturingPhase.call(this._parentNode, event);
            }
            if (_$jscmd("lib/Node.js", "cond", "71_10_24", event.propagationStopped)) {
                _$jscmd("lib/Node.js", "line", 72);
                return;
            }
            _$jscmd("lib/Node.js", "line", 74);
            var callbacks = _$jscmd("lib/Node.js", "cond", "74_22_26", this._eventsCapturingPhase) && _$jscmd("lib/Node.js", "cond", "74_52_42", this._eventsCapturingPhase.get(event.type));
            if (_$jscmd("lib/Node.js", "cond", "75_10_9", callbacks)) {
                _$jscmd("lib/Node.js", "line", 76);
                callbacks.some(function(callback) {
                    _$jscmd("lib/Node.js", "line", 77);
                    callback();
                    _$jscmd("lib/Node.js", "line", 78);
                    return event.immediatePropagationStopped;
                });
            }
        };
        _$jscmd("lib/Node.js", "line", 82);
        capturingPhase.call(this, event);
        // Bubbling phase
        if (_$jscmd("lib/Node.js", "cond", "85_8_25", !event.propagationStopped)) {
            _$jscmd("lib/Node.js", "line", 86);
            var bubblingPhase = function(event) {
                _$jscmd("lib/Node.js", "line", 87);
                var callbacks = _$jscmd("lib/Node.js", "cond", "87_24_25", this._eventsBubblingPhase) && _$jscmd("lib/Node.js", "cond", "87_53_41", this._eventsBubblingPhase.get(event.type));
                if (_$jscmd("lib/Node.js", "cond", "88_12_9", callbacks)) {
                    _$jscmd("lib/Node.js", "line", 89);
                    callbacks.some(function(callback) {
                        _$jscmd("lib/Node.js", "line", 90);
                        callback();
                        _$jscmd("lib/Node.js", "line", 91);
                        return event.immediatePropagationStopped;
                    });
                }
                if (_$jscmd("lib/Node.js", "cond", "94_12_45", _$jscmd("lib/Node.js", "cond", "94_12_25", !event.propagationStopped) && _$jscmd("lib/Node.js", "cond", "94_41_16", this._parentNode))) {
                    _$jscmd("lib/Node.js", "line", 95);
                    bubblingPhase.call(this._parentNode, event);
                }
            };
            _$jscmd("lib/Node.js", "line", 98);
            bubblingPhase.call(this, event);
        }
        _$jscmd("lib/Node.js", "line", 101);
        return !event.defaultPrevented;
    };
    _$jscmd("lib/Node.js", "line", 104);
    _classProps(Node, null, {
        textContent: {
            /* CONTENT */
            /**
       * @return {String}
       */
            get: function() {
                _$jscmd("lib/Node.js", "line", 112);
                return "";
            }
        }
    });
    _$jscmd("lib/Node.js", "line", 117);
    return Node;
}();

_$jscmd("lib/Node.js", "line", 120);

exports.Node = Node;

_$jscmd("lib/Node.js", "line", 123);

Object.defineProperty(Node, "ELEMENT_NODE", {
    value: 1
});

_$jscmd("lib/Node.js", "line", 124);

Object.defineProperty(Node, "ATTRIBUTE_NODE", {
    value: 2
});

_$jscmd("lib/Node.js", "line", 125);

Object.defineProperty(Node, "TEXT_NODE", {
    value: 3
});

_$jscmd("lib/Node.js", "line", 126);

Object.defineProperty(Node, "CDATA_SECTION_NODE", {
    value: 4
});

_$jscmd("lib/Node.js", "line", 127);

Object.defineProperty(Node, "ENTITY_REFERENCE_NODE", {
    value: 5
});

_$jscmd("lib/Node.js", "line", 128);

Object.defineProperty(Node, "ENTITY_NODE", {
    value: 6
});

_$jscmd("lib/Node.js", "line", 129);

Object.defineProperty(Node, "PROCESSING_INSTRUCTION_NODE", {
    value: 7
});

_$jscmd("lib/Node.js", "line", 130);

Object.defineProperty(Node, "COMMENT_NODE", {
    value: 8
});

_$jscmd("lib/Node.js", "line", 131);

Object.defineProperty(Node, "DOCUMENT_NODE", {
    value: 9
});

_$jscmd("lib/Node.js", "line", 132);

Object.defineProperty(Node, "DOCUMENT_TYPE_NODE", {
    value: 10
});

_$jscmd("lib/Node.js", "line", 133);

Object.defineProperty(Node, "DOCUMENT_FRAGMENT_NODE", {
    value: 11
});

_$jscmd("lib/Node.js", "line", 134);

Object.defineProperty(Node, "NOTATION_NODE", {
    value: 12
});
