// html-document/lib-cov/ParentNode.js

// instrument by jscoverage, do not modifly this file
(function (file, lines, conds, source) {
  var BASE;
  if (typeof global === 'object') {
    BASE = global;
  } else if (typeof window === 'object') {
    BASE = window;
  } else {
    throw new Error('[jscoverage] unknow ENV!');
  }
  if (BASE._$jscoverage) {
    BASE._$jscmd(file, 'init', lines, conds, source);
    return;
  }
  var cov = {};
  /**
   * jsc(file, 'init', lines, condtions)
   * jsc(file, 'line', lineNum)
   * jsc(file, 'cond', lineNum, expr, start, offset)
   */
  function jscmd(file, type, line, express, start, offset) {
    var storage;
    switch (type) {
      case 'init':
        if(cov[file]){
          storage = cov[file];
        } else {
          storage = [];
          for (var i = 0; i < line.length; i ++) {
            storage[line[i]] = 0;
          }
          var condition = express;
          var source = start;
          storage.condition = condition;
          storage.source = source;
        }
        cov[file] = storage;
        break;
      case 'line':
        storage = cov[file];
        storage[line] ++;
        break;
      case 'cond':
        storage = cov[file];
        storage.condition[line] ++;
        return express;
    }
  }

  BASE._$jscoverage = cov;
  BASE._$jscmd = jscmd;
  jscmd(file, 'init', lines, conds, source);
})('lib/ParentNode.js', [1,3,8,20,21,23,220,9,17,24,29,31,40,57,77,101,111,134,217,25,26,32,33,35,41,42,54,44,45,48,50,72,73,74,59,62,65,70,67,68,78,97,98,80,83,86,92,88,89,90,94,95,102,106,107,108,104,112,131,114,117,120,126,122,123,124,128,129,140,148,156,164,172,176,174,184,188,186,196,197,201,202,210,211], {"4_6_11":0,"5_6_13":0,"34_10_26":0,"43_10_14":0,"47_10_26":0,"49_12_20":0,"58_8_22":0,"61_8_17":0,"64_8_46":0,"79_8_12":0,"82_8_20":0,"85_8_49":0,"103_8_12":0,"113_8_12":0,"116_8_17":0,"119_8_46":0,"148_15_16":0,"156_15_19":0,"164_15_45":0,"173_12_20":0,"176_15_69":0,"176_15_19":0,"176_38_46":0,"185_12_20":0,"188_15_46":0}, ["\"use strict\";","","var _classProps = function (child, staticProps, instanceProps) {","  if (staticProps) Object.defineProperties(child, staticProps);","  if (instanceProps) Object.defineProperties(child.prototype, instanceProps);","};","","var _extends = function (child, parent) {","  child.prototype = Object.create(parent.prototype, {","    constructor: {","      value: child,","      enumerable: false,","      writable: true,","      configurable: true","    }","  });","  child.__proto__ = parent;","};","","var Node = require(\"./Node\").Node;","var parser = require(\"./parser\");","","var ParentNode = (function (Node) {","  var ParentNode = function ParentNode() {","    Node.call(this);","    this._childNodes = [];","  };","","  _extends(ParentNode, Node);","","  ParentNode.prototype._childNodesRecursiveForEach = function (callback) {","    this._childNodes.forEach(function (node) {","      callback(node);","      if (node instanceof ParentNode) {","        node._childNodesRecursiveForEach(callback);","      }","    });","  };","","  ParentNode.prototype._childNodesRecursiveFind = function (callback) {","    var result;","    this._childNodes.some(function (node) {","      if (callback(node)) {","        result = node;","        return true;","      }","      if (node instanceof ParentNode) {","        result = node._childNodesRecursiveFind(callback);","        if (result !== undefined) {","          return true;","        }","      }","    });","    return result;","  };","","  ParentNode.prototype.appendChild = function (child) {","    if (!child instanceof Node) {","      throw new Error(\"Trying to add non node element\");","    }","    if (child._parentNode) {","      child._parentNode.removeChild(child);","    }","    if (child.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {","      var childNode;","      while (childNode = child.firstChild) {","        child.removeChild(childNode);","        this.appendChild(childNode);","      }","      return child;","    }","    child._parentNode = this;","    this._childNodes.push(child);","    return child;","  };","","  ParentNode.prototype.replaceChild = function (newChild, oldChild) {","    var index = this._childNodes.indexOf(oldChild);","    if (index === -1) {","      throw new Error(\"Node was not found\");","    }","    if (newChild._parentNode) {","      newChild._parentNode.removeChild(newChild);","    }","    if (newChild.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {","      var newChildren = [], childNode;","      while (childNode = newChild.firstChild) {","        newChild.removeChild(childNode);","        childNode._parentNode = this;","        newChildren.push(childNode);","      }","      this._childNodes.splice.bind(this._childNodes, index, 1).apply(null, newChildren);","    } else {","      newChild._parentNode = this;","      this._childNodes[index] = newChild;","    }","    delete oldChild._parentNode;","    return oldChild;","  };","","  ParentNode.prototype.removeChild = function (toRemoveChild) {","    var index = this._childNodes.indexOf(toRemoveChild);","    if (index === -1) {","      throw new Error(\"Node was not found\");","    }","    this._childNodes.splice(index, 1);","    delete toRemoveChild._parentNode;","    return toRemoveChild;","  };","","  ParentNode.prototype.insertBefore = function (child, existingChild) {","    var index = this._childNodes.indexOf(existingChild);","    if (index === -1) {","      throw new Error(\"Node was not found\");","    }","    if (child._parentNode) {","      child._parentNode.removeChild(child);","    }","    if (child.nodeType === Node.DOCUMENT_FRAGMENT_NODE) {","      var children = [], childNode;","      while (childNode = child.firstChild) {","        child.removeChild(childNode);","        childNode._parentNode = this;","        children.push(childNode);","      }","      this._childNodes.splice.bind(this._childNodes, index, 0).apply(null, children);","    } else {","      child._parentNode = this;","      this._childNodes.splice(index, 0, child);","    }","    return child;","  };","","  _classProps(ParentNode, null, {","    childNodes: {","      /**","       * @return {HTMLCollection}","       */","      get: function () {","        return this._childNodes;","      }","    },","    parentNode: {","      /**","       * @return {Node|null}","       */","      get: function () {","        return this._parentNode || null;","      }","    },","    firstChild: {","      /**","       * @return {Node|null}","       */","      get: function () {","        return this._childNodes[0] || null;","      }","    },","    lastChild: {","      /**","       * @return {Node|null}","       */","      get: function () {","        return this._childNodes[this._childNodes.length - 1] || null;","      }","    },","    previousSibling: {","      /**","       * @return {Node|null}","       */","      get: function () {","        var indexInParent = this.parentNode._childNodes.indexOf(this);","        if (indexInParent === -1) {","          throw new Error(\"Unexpected state: this node is not in the parent\");","        }","        return indexInParent !== 0 && this.parentNode._childNodes[indexInParent - 1] || null;","      }","    },","    nextSibling: {","      /**","       * @return {Node|null}","       */","      get: function () {","        var indexInParent = this.parentNode._childNodes.indexOf(this);","        if (indexInParent === -1) {","          throw new Error(\"Unexpected state: this node is not in the parent\");","        }","        return this.parentNode._childNodes[indexInParent + 1] || null;","      }","    },","    innerHTML: {","      /**","       * @return {String}","       */","      get: function () {","        return this._childNodes.reduce(function (value, node) {","          return value + node._toHTML();","        }, \"\");","      },","      set: function (html) {","        this._childNodes = [];","        parser.parse(html, this);","      }","    },","    textContent: {","      /**","       * @return {String}","       */","      get: function () {","        return this._childNodes.reduce(function (value, node) {","          return value + node.textContent;","        }, \"\");","      }","    }","  });","","  return ParentNode;","})(Node);","","exports.ParentNode = ParentNode;","//# sourceMappingURL=ParentNode.js.map"]);
_$jscmd("lib/ParentNode.js", "line", 1);

"use strict";

_$jscmd("lib/ParentNode.js", "line", 3);

var _classProps = function(child, staticProps, instanceProps) {
    if (_$jscmd("lib/ParentNode.js", "cond", "4_6_11", staticProps)) Object.defineProperties(child, staticProps);
    if (_$jscmd("lib/ParentNode.js", "cond", "5_6_13", instanceProps)) Object.defineProperties(child.prototype, instanceProps);
};

_$jscmd("lib/ParentNode.js", "line", 8);

var _extends = function(child, parent) {
    _$jscmd("lib/ParentNode.js", "line", 9);
    child.prototype = Object.create(parent.prototype, {
        constructor: {
            value: child,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    _$jscmd("lib/ParentNode.js", "line", 17);
    child.__proto__ = parent;
};

_$jscmd("lib/ParentNode.js", "line", 20);

var Node = require("./Node").Node;

_$jscmd("lib/ParentNode.js", "line", 21);

var parser = require("./parser");

_$jscmd("lib/ParentNode.js", "line", 23);

var ParentNode = function(Node) {
    _$jscmd("lib/ParentNode.js", "line", 24);
    var ParentNode = function ParentNode() {
        _$jscmd("lib/ParentNode.js", "line", 25);
        Node.call(this);
        _$jscmd("lib/ParentNode.js", "line", 26);
        this._childNodes = [];
    };
    _$jscmd("lib/ParentNode.js", "line", 29);
    _extends(ParentNode, Node);
    _$jscmd("lib/ParentNode.js", "line", 31);
    ParentNode.prototype._childNodesRecursiveForEach = function(callback) {
        _$jscmd("lib/ParentNode.js", "line", 32);
        this._childNodes.forEach(function(node) {
            _$jscmd("lib/ParentNode.js", "line", 33);
            callback(node);
            if (_$jscmd("lib/ParentNode.js", "cond", "34_10_26", node instanceof ParentNode)) {
                _$jscmd("lib/ParentNode.js", "line", 35);
                node._childNodesRecursiveForEach(callback);
            }
        });
    };
    _$jscmd("lib/ParentNode.js", "line", 40);
    ParentNode.prototype._childNodesRecursiveFind = function(callback) {
        _$jscmd("lib/ParentNode.js", "line", 41);
        var result;
        _$jscmd("lib/ParentNode.js", "line", 42);
        this._childNodes.some(function(node) {
            if (_$jscmd("lib/ParentNode.js", "cond", "43_10_14", callback(node))) {
                _$jscmd("lib/ParentNode.js", "line", 44);
                result = node;
                _$jscmd("lib/ParentNode.js", "line", 45);
                return true;
            }
            if (_$jscmd("lib/ParentNode.js", "cond", "47_10_26", node instanceof ParentNode)) {
                _$jscmd("lib/ParentNode.js", "line", 48);
                result = node._childNodesRecursiveFind(callback);
                if (_$jscmd("lib/ParentNode.js", "cond", "49_12_20", result !== undefined)) {
                    _$jscmd("lib/ParentNode.js", "line", 50);
                    return true;
                }
            }
        });
        _$jscmd("lib/ParentNode.js", "line", 54);
        return result;
    };
    _$jscmd("lib/ParentNode.js", "line", 57);
    ParentNode.prototype.appendChild = function(child) {
        if (_$jscmd("lib/ParentNode.js", "cond", "58_8_22", !child instanceof Node)) {
            _$jscmd("lib/ParentNode.js", "line", 59);
            throw new Error("Trying to add non node element");
        }
        if (_$jscmd("lib/ParentNode.js", "cond", "61_8_17", child._parentNode)) {
            _$jscmd("lib/ParentNode.js", "line", 62);
            child._parentNode.removeChild(child);
        }
        if (_$jscmd("lib/ParentNode.js", "cond", "64_8_46", child.nodeType === Node.DOCUMENT_FRAGMENT_NODE)) {
            _$jscmd("lib/ParentNode.js", "line", 65);
            var childNode;
            while (childNode = child.firstChild) {
                _$jscmd("lib/ParentNode.js", "line", 67);
                child.removeChild(childNode);
                _$jscmd("lib/ParentNode.js", "line", 68);
                this.appendChild(childNode);
            }
            _$jscmd("lib/ParentNode.js", "line", 70);
            return child;
        }
        _$jscmd("lib/ParentNode.js", "line", 72);
        child._parentNode = this;
        _$jscmd("lib/ParentNode.js", "line", 73);
        this._childNodes.push(child);
        _$jscmd("lib/ParentNode.js", "line", 74);
        return child;
    };
    _$jscmd("lib/ParentNode.js", "line", 77);
    ParentNode.prototype.replaceChild = function(newChild, oldChild) {
        _$jscmd("lib/ParentNode.js", "line", 78);
        var index = this._childNodes.indexOf(oldChild);
        if (_$jscmd("lib/ParentNode.js", "cond", "79_8_12", index === -1)) {
            _$jscmd("lib/ParentNode.js", "line", 80);
            throw new Error("Node was not found");
        }
        if (_$jscmd("lib/ParentNode.js", "cond", "82_8_20", newChild._parentNode)) {
            _$jscmd("lib/ParentNode.js", "line", 83);
            newChild._parentNode.removeChild(newChild);
        }
        if (_$jscmd("lib/ParentNode.js", "cond", "85_8_49", newChild.nodeType === Node.DOCUMENT_FRAGMENT_NODE)) {
            _$jscmd("lib/ParentNode.js", "line", 86);
            var newChildren = [], childNode;
            while (childNode = newChild.firstChild) {
                _$jscmd("lib/ParentNode.js", "line", 88);
                newChild.removeChild(childNode);
                _$jscmd("lib/ParentNode.js", "line", 89);
                childNode._parentNode = this;
                _$jscmd("lib/ParentNode.js", "line", 90);
                newChildren.push(childNode);
            }
            _$jscmd("lib/ParentNode.js", "line", 92);
            this._childNodes.splice.bind(this._childNodes, index, 1).apply(null, newChildren);
        } else {
            _$jscmd("lib/ParentNode.js", "line", 94);
            newChild._parentNode = this;
            _$jscmd("lib/ParentNode.js", "line", 95);
            this._childNodes[index] = newChild;
        }
        _$jscmd("lib/ParentNode.js", "line", 97);
        delete oldChild._parentNode;
        _$jscmd("lib/ParentNode.js", "line", 98);
        return oldChild;
    };
    _$jscmd("lib/ParentNode.js", "line", 101);
    ParentNode.prototype.removeChild = function(toRemoveChild) {
        _$jscmd("lib/ParentNode.js", "line", 102);
        var index = this._childNodes.indexOf(toRemoveChild);
        if (_$jscmd("lib/ParentNode.js", "cond", "103_8_12", index === -1)) {
            _$jscmd("lib/ParentNode.js", "line", 104);
            throw new Error("Node was not found");
        }
        _$jscmd("lib/ParentNode.js", "line", 106);
        this._childNodes.splice(index, 1);
        _$jscmd("lib/ParentNode.js", "line", 107);
        delete toRemoveChild._parentNode;
        _$jscmd("lib/ParentNode.js", "line", 108);
        return toRemoveChild;
    };
    _$jscmd("lib/ParentNode.js", "line", 111);
    ParentNode.prototype.insertBefore = function(child, existingChild) {
        _$jscmd("lib/ParentNode.js", "line", 112);
        var index = this._childNodes.indexOf(existingChild);
        if (_$jscmd("lib/ParentNode.js", "cond", "113_8_12", index === -1)) {
            _$jscmd("lib/ParentNode.js", "line", 114);
            throw new Error("Node was not found");
        }
        if (_$jscmd("lib/ParentNode.js", "cond", "116_8_17", child._parentNode)) {
            _$jscmd("lib/ParentNode.js", "line", 117);
            child._parentNode.removeChild(child);
        }
        if (_$jscmd("lib/ParentNode.js", "cond", "119_8_46", child.nodeType === Node.DOCUMENT_FRAGMENT_NODE)) {
            _$jscmd("lib/ParentNode.js", "line", 120);
            var children = [], childNode;
            while (childNode = child.firstChild) {
                _$jscmd("lib/ParentNode.js", "line", 122);
                child.removeChild(childNode);
                _$jscmd("lib/ParentNode.js", "line", 123);
                childNode._parentNode = this;
                _$jscmd("lib/ParentNode.js", "line", 124);
                children.push(childNode);
            }
            _$jscmd("lib/ParentNode.js", "line", 126);
            this._childNodes.splice.bind(this._childNodes, index, 0).apply(null, children);
        } else {
            _$jscmd("lib/ParentNode.js", "line", 128);
            child._parentNode = this;
            _$jscmd("lib/ParentNode.js", "line", 129);
            this._childNodes.splice(index, 0, child);
        }
        _$jscmd("lib/ParentNode.js", "line", 131);
        return child;
    };
    _$jscmd("lib/ParentNode.js", "line", 134);
    _classProps(ParentNode, null, {
        childNodes: {
            /**
       * @return {HTMLCollection}
       */
            get: function() {
                _$jscmd("lib/ParentNode.js", "line", 140);
                return this._childNodes;
            }
        },
        parentNode: {
            /**
       * @return {Node|null}
       */
            get: function() {
                _$jscmd("lib/ParentNode.js", "line", 148);
                return _$jscmd("lib/ParentNode.js", "cond", "148_15_16", this._parentNode) || null;
            }
        },
        firstChild: {
            /**
       * @return {Node|null}
       */
            get: function() {
                _$jscmd("lib/ParentNode.js", "line", 156);
                return _$jscmd("lib/ParentNode.js", "cond", "156_15_19", this._childNodes[0]) || null;
            }
        },
        lastChild: {
            /**
       * @return {Node|null}
       */
            get: function() {
                _$jscmd("lib/ParentNode.js", "line", 164);
                return _$jscmd("lib/ParentNode.js", "cond", "164_15_45", this._childNodes[this._childNodes.length - 1]) || null;
            }
        },
        previousSibling: {
            /**
       * @return {Node|null}
       */
            get: function() {
                _$jscmd("lib/ParentNode.js", "line", 172);
                var indexInParent = this.parentNode._childNodes.indexOf(this);
                if (_$jscmd("lib/ParentNode.js", "cond", "173_12_20", indexInParent === -1)) {
                    _$jscmd("lib/ParentNode.js", "line", 174);
                    throw new Error("Unexpected state: this node is not in the parent");
                }
                _$jscmd("lib/ParentNode.js", "line", 176);
                return _$jscmd("lib/ParentNode.js", "cond", "176_15_69", _$jscmd("lib/ParentNode.js", "cond", "176_15_19", indexInParent !== 0) && _$jscmd("lib/ParentNode.js", "cond", "176_38_46", this.parentNode._childNodes[indexInParent - 1])) || null;
            }
        },
        nextSibling: {
            /**
       * @return {Node|null}
       */
            get: function() {
                _$jscmd("lib/ParentNode.js", "line", 184);
                var indexInParent = this.parentNode._childNodes.indexOf(this);
                if (_$jscmd("lib/ParentNode.js", "cond", "185_12_20", indexInParent === -1)) {
                    _$jscmd("lib/ParentNode.js", "line", 186);
                    throw new Error("Unexpected state: this node is not in the parent");
                }
                _$jscmd("lib/ParentNode.js", "line", 188);
                return _$jscmd("lib/ParentNode.js", "cond", "188_15_46", this.parentNode._childNodes[indexInParent + 1]) || null;
            }
        },
        innerHTML: {
            /**
       * @return {String}
       */
            get: function() {
                _$jscmd("lib/ParentNode.js", "line", 196);
                return this._childNodes.reduce(function(value, node) {
                    _$jscmd("lib/ParentNode.js", "line", 197);
                    return value + node._toHTML();
                }, "");
            },
            set: function(html) {
                _$jscmd("lib/ParentNode.js", "line", 201);
                this._childNodes = [];
                _$jscmd("lib/ParentNode.js", "line", 202);
                parser.parse(html, this);
            }
        },
        textContent: {
            /**
       * @return {String}
       */
            get: function() {
                _$jscmd("lib/ParentNode.js", "line", 210);
                return this._childNodes.reduce(function(value, node) {
                    _$jscmd("lib/ParentNode.js", "line", 211);
                    return value + node.textContent;
                }, "");
            }
        }
    });
    _$jscmd("lib/ParentNode.js", "line", 217);
    return ParentNode;
}(Node);

_$jscmd("lib/ParentNode.js", "line", 220);

exports.ParentNode = ParentNode;
